### PR TITLE
[SHELL32] Read the label for a CD from autorun.inf

### PR DESCRIPTION
**Pull request by Cleverking2003**
## Purpose

Some CDs may set a custom label using autorun.inf. Read the label from the file, and if it succeeds, show it to the user.

JIRA issue: [CORE-18567](https://jira.reactos.org/browse/CORE-18567)

## Proposed changes

Add a function to read the `label` variable from autorun.inf and use it.

image

### CODE REVIEW

**Comment by JoachimHenze** [comment]
I do remember that MS decided to stop some of the processing within autorun.infs from external media for security considerations in newer versions of Windows. Especially the job of actually performing the "autorun". Just reading the label doesn't sound too dangerous to me.
Is no longer reading the label part of that hardening or not? In other words:
Does Windows **10** still read those labels from autorun.inf on CD drives by default?

**Comment by Cleverking2003** [comment]
> I do remember that MS decided to stop some of the processing within autorun.infs from external media for security considerations in newer versions of Windows. Especially the job of actually performing the "autorun". Just reading the label doesn't sound too dangerous to me. Is no longer reading the label part of that hardening or not? In other words: Does Windows **10** still read those labels from autorun.inf on CD drives by default?

Windows 10 still shows the label, I think that the restrictions only affect the actual running of programs from the disk.

File: dll/win32/shell32/folders/CDrivesFolder.cpp [hunk 1]
```diff
@@ -979,38 +1001,42 @@ HRESULT WINAPI CDrivesFolder::GetDisplayNameOf(PCUITEMID_CHILD pidl, DWORD dwFla
...
-                    break;
-                case DRIVE_CDROM:
-                    ResourceId = IDS_DRIVE_CDROM;
-                    break;
-                default:
-                    ResourceId = 0;
-            }
-            if (ResourceId)
+
+        if (!SUCCEEDED(getLabelForDrive(wszDrive, pszPath))) {
```
Comment (gonzoMD) [inline dll/win32/shell32/folders/CDrivesFolder.cpp:1008]: Please move the curly brace to its own line

Comment (Cleverking2003) [reply dll/win32/shell32/folders/CDrivesFolder.cpp]: Oops, fixed

**Review by JoachimHenze — APPROVED** [review]
Approved, since you did just move (and not created) the code piece that makes me uncomfortable most: What is MAX_PATH - 7 ?

**Comment by Cleverking2003** [comment]
> Approved, since you did just move (and not created) the code piece that makes me uncomfortable most: What is MAX_PATH - 7 ?

Probably it leaves 7 characters for the ` (X:)` part (though that only has 6 characters together with `\0`)

**Review by learn-more — APPROVED** [review]
While I am not very happy with the fixed size buffers for paths, it is following the style of the surrounding code.

File: dll/win32/shell32/folders/CDrivesFolder.cpp [hunk 2]
```diff
@@ -979,38 +1001,43 @@ HRESULT WINAPI CDrivesFolder::GetDisplayNameOf(PCUITEMID_CHILD pidl, DWORD dwFla
...
-            DriveType = GetDriveTypeW(wszDrive);
-            switch(DriveType)
+            DWORD dwVolumeSerialNumber, dwMaximumComponentLength, dwFileSystemFlags;
+
+            GetVolumeInformationW(wszDrive, pszPath,
+                                    MAX_PATH - 7,
+                                    &dwVolumeSerialNumber,
+                                    &dwMaximumComponentLength, &dwFileSystemFlags, NULL, 0);
+            pszPath[MAX_PATH-1] = L'\0';
+            if (!wcslen(pszPath))
```
Comment (binarymaster) [inline dll/win32/shell32/folders/CDrivesFolder.cpp:1017]:
```suggestion

            if (!wcslen(pszPath))
```

File: dll/win32/shell32/folders/CDrivesFolder.cpp [hunk 3]
```diff
@@ -979,38 +1001,43 @@ HRESULT WINAPI CDrivesFolder::GetDisplayNameOf(PCUITEMID_CHILD pidl, DWORD dwFla
...
-                    ResourceId = 0;
-            }
-            if (ResourceId)
-            {
-                dwFileSystemFlags = LoadStringW(shell32_hInstance, ResourceId, pszPath, MAX_PATH);
-                if (dwFileSystemFlags > MAX_PATH - 7)
-                    pszPath[MAX_PATH-7] = L'\0';
+                UINT DriveType, ResourceId;
+                DriveType = GetDriveTypeW(wszDrive);
+                switch(DriveType)
```
Comment (binarymaster) [inline dll/win32/shell32/folders/CDrivesFolder.cpp:1021]:
```suggestion

                switch (DriveType)
```

File: dll/win32/shell32/folders/CDrivesFolder.cpp [hunk 4]
```diff
@@ -979,38 +1001,43 @@ HRESULT WINAPI CDrivesFolder::GetDisplayNameOf(PCUITEMID_CHILD pidl, DWORD dwFla
...
+                    case DRIVE_REMOTE:
+                        ResourceId = IDS_DRIVE_NETWORK;
+                        break;
+                    case DRIVE_CDROM:
+                        ResourceId = IDS_DRIVE_CDROM;
+                        break;
+                    default:
+                        ResourceId = 0;
+                }
+                if (ResourceId)
```
Comment (binarymaster) [inline dll/win32/shell32/folders/CDrivesFolder.cpp:1035]:
```suggestion

                if (ResourceId)
```

**Review by binarymaster — COMMENTED** [review]
Some spacing for better readability.

**Review by binarymaster — APPROVED** [review]

**Pull request merged**